packages/rpc-transport/src/transports/websocket/__tests__/websocket-connection-test.ts

```diff
@@ -0,0 +1,114 @@
+import WS from 'jest-websocket-mock';
+import { Client } from 'mock-socket';
+
+import { createWebSocketConnection, RpcWebSocketConnection } from '../websocket-connection';
+
+describe('RpcWebSocketConnection', () => {
+    let abortController: AbortController;
+    let connection: RpcWebSocketConnection;
+    let ws: WS;
+    function getLatestClient() {
+        const clients = ws.server.clients();
+        return clients[clients.length - 1];
+    }
+    beforeEach(async () => {
+        abortController = new AbortController();
+        ws = new WS('wss://fake', {
+            jsonProtocol: true,
+        });
+        connection = await createWebSocketConnection({
+            signal: abortController.signal,
+            url: 'wss://fake',
+        });
+        await ws.connected;
+    });
+    afterEach(() => {
+        WS.clean();
+    });
+    it('does not resolve until the socket is open', async () => {
+        expect.assertions(2);
+        const freshConnectionPromise = createWebSocketConnection({
+            signal: abortController.signal,
+            url: 'wss://fake',
+        });
+        const client = getLatestClient();
+        expect(client).toHaveProperty('readyState', WebSocket.CONNECTING);
+        await freshConnectionPromise;
+        expect(client).toHaveProperty('readyState', WebSocket.OPEN);
+    });
+    it('vends a message to consumers who have already polled for a result', async () => {
+        expect.assertions(2);
+        const iteratorA = connection[Symbol.asyncIterator]();
+        const iteratorB = connection[Symbol.asyncIterator]();
+        const resultPromiseA = iteratorA.next();
+        const resultPromiseB = iteratorB.next();
+        const expectedMessage = { some: 'message' };
+        ws.send(expectedMessage);
+        await expect(resultPromiseA).resolves.toMatchObject({ done: false, value: expectedMessage });
+        await expect(resultPromiseB).resolves.toMatchObject({ done: false, value: expectedMessage });
+    });
+    it('does not queue messsages for a consumer until it has started to poll', async () => {
+        expect.assertions(3);
+        const iterator = connection[Symbol.asyncIterator]();
+        ws.send({ some: 'lost message' });
+        const resultPromise = iterator.next();
+        ws.send({ some: 'immediately delivered message' });
+        await expect(resultPromise).resolves.toMatchObject({
+            done: false,
+            value: { some: 'immediately delivered message' },
+        });
+        ws.send({ some: 'queued message 1' });
+        ws.send({ some: 'queued message 2' });
+        await expect(iterator.next()).resolves.toMatchObject({
+            done: false,
+            value: { some: 'queued message 1' },
+        });
+        await expect(iterator.next()).resolves.toMatchObject({
+            done: false,
+            value: { some: 'queued message 2' },
+        });
```

Comment:
Yeah it's pretty insane how you can use an iterator like this:

```ts
try {
  for await (const nextItem of stuff) {
    console.log('I called `next()` on `stuff` and all I got was', nextItem);
  }
  // If you end up here it's because something called `return()` on `stuff`
} catch(e) {
  // If you end up here it's because something called `throw()` on `stuff`
}
```

…but you can also sort of manually crank it by calling `.next()`, `.return()`, or `.throw()`.

Makes it really easy to write tests.